lib/services/local/dynamic_access.go

```diff
@@ -158,6 +176,138 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
 	return nil
 }
 
+// GetPluginData loads all plugin data matching the supplied filter.
+func (s *DynamicAccessService) GetPluginData(ctx context.Context, filter services.PluginDataFilter) ([]services.PluginData, error) {
+	switch filter.Kind {
+	case services.KindAccessRequest:
+		data, err := s.getAccessRequestPluginData(ctx, filter)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return data, nil
+	default:
+		return nil, trace.BadParameter("unsupported resource kind %q", filter.Kind)
+	}
+}
+
+func (s *DynamicAccessService) getAccessRequestPluginData(ctx context.Context, filter services.PluginDataFilter) ([]services.PluginData, error) {
+	// Filters which specify Resource are a special case since they will match exactly zero or one
+	// possible PluginData instances.
+	if filter.Resource != "" {
+		item, err := s.Get(ctx, pluginDataKey(services.KindAccessRequest, filter.Resource))
+		if err != nil {
+			if trace.IsNotFound(err) {
+				// Caller is going to need to check the length of the returned slice anyway,
+				// so propagating a NotFoundError is redundant.
+				return nil, nil
+			}
+			return nil, trace.Wrap(err)
+		}
+		data, err := itemToPluginData(*item)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if !filter.Match(data) {
+			return nil, nil
+		}
+		return []services.PluginData{data}, nil
+	}
+	prefix := backend.Key(pluginDataPrefix, services.KindAccessRequest)
+	result, err := s.GetRange(ctx, prefix, backend.RangeEnd(prefix), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var matches []services.PluginData
+	for _, item := range result.Items {
+		if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
+			// Item represents a different resource type in the
+			// same namespace.
+			continue
+		}
+		data, err := itemToPluginData(item)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if !filter.Match(data) {
+			continue
+		}
+		matches = append(matches, data)
+	}
+	return matches, nil
+}
+
+// UpdatePluginData updates a per-resource PluginData entry.
+func (s *DynamicAccessService) UpdatePluginData(ctx context.Context, params services.PluginDataUpdateParams) error {
+	switch params.Kind {
+	case services.KindAccessRequest:
+		return trace.Wrap(s.updateAccessRequestPluginData(ctx, params))
+	default:
+		return trace.BadParameter("unsupported resource kind %q", params.Kind)
+	}
+}
+
+func (s *DynamicAccessService) updateAccessRequestPluginData(ctx context.Context, params services.PluginDataUpdateParams) error {
+	// Update is attempted multiple times in the event of concurrent writes.
+	for i := 0; i < maxCmpAttempts; i++ {
```

Comment:
same here, with no backoff this will generate a spin lock with servers consuming 100% CPU on throttling or service outage.